### ISSUE TITLE
Unsound handling of unions with multiple type variables

### ISSUE THREAD

**Issue by mheiber** [microsoft/TypeScript#49186]
# Bug Report

### 🔎 Search Terms

unsound,union,generic

### 🕗 Version & Regression Information

- This is a runtime error
- This is the behavior in every version I tried, and I reviewed the FAQ for entries about: **unions**

### ⏯ Playground Link

[playground link](https://www.typescriptlang.org/play?ts=3.3.3#code/GYVwdgxgLglg9mABAUwB5QE4ENoB4DSAfABSoBciAghtgJ4GGIA+VNW91dDhAlBfogDeAKESIICAM5REWGogC8iVAG0ADAF0A3KMQxgiYp3YA6GJOO1icjDx5DdYjMiggMSG+o26Avrueu7rI0wn7CEmDSyooo6NjQxCoqAMwaGvYA9BmIABZwAG7IGIhQcIiSyMgltAAOyBQABmAgALYARkUqGg3hUnAANsgm-XAA5qSZ2W0gMvlY-SBV5ogNyT2oJi1YNaSKjKiTiADuOTAQOYiDWAAmkiVlGOCwLVVFGHAYFADkG1s1encwHAZFhEKBILAEF9hEA)

### 💻 Code

```ts
function extract<K>(x: Array<K> | Array<Array<K>>): K {
  const arr = x[0];
  if (Array.isArray(arr)) {
    return arr[0]
  }
  return arr
}

const x = extract([[3]]) // hover to see type: `number[]`
console.log(x) // but value is `3`
x.map(x => x) // which leads to runtime error: 'x.map is not a function'
```

### 🙁 Actual behavior

Runtime error not caught by TypeScript, due to incorrect inference.

### 🙂 Expected behavior

Type error.

Could be in any of these places:
- in `extract`, since the type signature is probably a mistake
- on the last line. Which could be achieved if the return type of `extract([[3]])` were `number | number[]`

**Comment by whzx5byb** [comment]
It works as expected after v3.6.2 in playground.

**Comment by MartinJohns** [comment]
> This is the behavior in every version I tried

Which versions have you tried?

**Comment by RyanCavanaugh** [comment]
> `v3.3.3333`

Doing some archaeology? 😉

**Comment by mheiber** [comment]
I've updated the code example, now there are no errors at type-checking time even on latest TS (4.7.0-beta, 4.6.4, 4.5.5, 4.4.4, 4.2.3, 4.1.5, 4.0.5, 3.9.7, 3.8.3, 3.7.5, 3.6.3, 3.5.1, and 3.3.3).

Sorry about that, what happened was I over-reduced the example when trying on older versions. Could we re-open?

**Comment by mheiber** [comment]
Update: opening as #49432. Sorry, there was a mistake in my previous repro
